// selecttransform/dist/key-executors/each.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.Each = void 0;
var internal_1 = require("../internal");
var Each = /** @class */ (function () {
    function Each() {
        this.name = Each.name;
    }
    Each.prototype.getName = function () {
        return this.name;
    };
    Each.prototype.fits = function (template) {
        return (typeof template === 'string' &&
            /^\s*\{\{\s*#each\s+.+\s*\}\}\s*$/g.test(template.toLowerCase()));
    };
    Each.prototype.execute = function (template, data, ts, key, result) {
        return __awaiter(this, void 0, void 0, function () {
            var fun, dataArray, err, promises, d, _loop_1, index;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        fun = internal_1.Helper.tokenize(key);
                        dataArray = internal_1.Helper.fillout('{{' + fun.expression + '}}', data, true);
                        // Ideally newData should be an array since it was prefixed by #each
                        if (!dataArray || !internal_1.Helper.isArray(dataArray)) {
                            err = internal_1.ST_ERRORS.data;
                            err.message += " - Wrong " + Each.name + " data - expected an array as the iterator.";
                            throw err;
                        }
                        promises = [];
                        d = JSON.parse(JSON.stringify(dataArray));
                        _loop_1 = function (index) {
                            var i = index;
                            promises.push(
                            // tslint:disable-next-line: ter-arrow-parens
                            new Promise(function (res) { return __awaiter(_this, void 0, void 0, function () {
                                var t, k, loopItem, k;
                                return __generator(this, function (_a) {
                                    switch (_a.label) {
                                        case 0:
                                            t = ts.copy();
                                            // temporarily set $index and $this
                                            if (typeof d[i] === 'object') {
                                                d[i]['$index'] = i;
                                                d[i]['$this'] = d[i];
                                                d[i]['$root'] = data['$root'];
                                                // Copy in the memory
                                                for (k in t.memory) {
                                                    // Only if we do not override anything
                                                    if (d[i][k] === undefined) {
                                                        d[i][k] = t.memory[k];
                                                    }
                                                }
                                            }
                                            return [4 /*yield*/, t.run(template[key], d[i])];
                                        case 1:
                                            loopItem = _a.sent();
                                            // clean up $index and $this
                                            if (typeof d[i] === 'object') {
                                                delete d[i]['$index'];
                                                delete d[i]['$this'];
                                                delete d[i]['$root'];
                                                // Copy in the memory
                                                // tslint:disable-next-line: forin
                                                for (k in t.memory) {
                                                    // Delete what we added
                                                    delete d[i][k];
                                                }
                                            }
                                            res(loopItem);
                                            return [2 /*return*/];
                                    }
                                });
                            }); }));
                        };
                        for (index = 0; index < dataArray.length; index++) {
                            _loop_1(index);
                        }
                        return [4 /*yield*/, Promise.all(promises)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Each.prototype.executeSync = function (template, data, ts, key, result) {
        var fun = internal_1.Helper.tokenize(key);
        // newData will be filled with parsed results
        var dataArray = internal_1.Helper.fillout('{{' + fun.expression + '}}', data, true);
        // Ideally newData should be an array since it was prefixed by #each
        if (!dataArray || !internal_1.Helper.isArray(dataArray)) {
            var err = internal_1.ST_ERRORS.data;
            err.message += " - Wrong " + Each.name + " data - expected an array as the iterator.";
            throw err;
        }
        result = [];
        for (var index = 0; index < dataArray.length; index++) {
            // temporarily set $index and $this
            if (typeof dataArray[index] === 'object') {
                dataArray[index]['$index'] = index;
                dataArray[index]['$this'] = dataArray[index];
                dataArray[index]['$root'] = data['$root'];
                // Copy in the memory
                for (var k in ts.memory) {
                    // Only if we do not override anything
                    if (dataArray[index][k] === undefined) {
                        dataArray[index][k] = ts.memory[k];
                    }
                }
            }
            else {
                String.prototype.$index = index;
                String.prototype.$this = dataArray[index];
                Number.prototype.$index = index;
                Number.prototype.$this = dataArray[index];
                Function.prototype.$index = index;
                Function.prototype.$this = dataArray[index];
                Array.prototype.$index = index;
                Array.prototype.$this = dataArray[index];
                Boolean.prototype.$index = index;
                Boolean.prototype.$this = dataArray[index];
            }
            // run
            var loopItem = ts.runSync(template[key], dataArray[index]);
            // clean up $index and $this
            if (typeof dataArray[index] === 'object') {
                delete dataArray[index]['$index'];
                delete dataArray[index]['$this'];
                delete dataArray[index]['$root'];
                // Copy in the memory
                // tslint:disable-next-line: forin
                for (var k in ts.memory) {
                    // Delete what we added
                    delete dataArray[index][k];
                }
            }
            else {
                delete String.prototype.$index;
                delete String.prototype.$this;
                delete Number.prototype.$index;
                delete Number.prototype.$this;
                delete Function.prototype.$index;
                delete Function.prototype.$this;
                delete Array.prototype.$index;
                delete Array.prototype.$this;
                delete Boolean.prototype.$index;
                delete Boolean.prototype.$this;
            }
            if (loopItem !== undefined && loopItem !== null) {
                // only push when the result is not null nor undefined
                // null could mean #if clauses where nothing matched => In this case
                // instead of rendering 'null', should just skip it completely
                result.push(loopItem);
            }
        }
        return result;
    };
    return Each;
}());
exports.Each = Each;
