// selecttransform/dist/key-executors/flatten.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
exports.Flatten = void 0;
var internal_1 = require("../internal");
var Flatten = /** @class */ (function () {
    function Flatten() {
        this.name = Flatten.name;
    }
    Flatten.prototype.getName = function () {
        return this.name;
    };
    Flatten.prototype.fits = function (template) {
        return (typeof template === 'string' &&
            /^\s*\{\{\s*#flatten\s*\}\}\s*$/g.test(template.toLowerCase()));
    };
    Flatten.prototype.executeSync = function (template, data, ts, key, result) {
        var obj = ts.runSync(template[key], data);
        result = [];
        if (internal_1.Helper.isArray(obj)) {
            var arr = obj;
            // For each item in the array
            for (var i = 0; i < arr.length; i++) {
                if (internal_1.Helper.isArray(arr[i])) {
                    // If array then flatten
                    for (var j = 0; j < arr[i].length; j++) {
                        result.push(arr[i][j]);
                    }
                }
                else {
                    // Just push if anything else
                    result.push(arr[i]);
                }
            }
        }
        return result;
    };
    Flatten.prototype.execute = function (template, data, ts, key, result) {
        return __awaiter(this, void 0, void 0, function () {
            var obj, arr, i, j;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, ts.run(template[key], data)];
                    case 1:
                        obj = _a.sent();
                        result = [];
                        if (internal_1.Helper.isArray(obj)) {
                            arr = obj;
                            // For each item in the array
                            for (i = 0; i < arr.length; i++) {
                                if (internal_1.Helper.isArray(arr[i])) {
                                    // If array then flatten
                                    for (j = 0; j < arr[i].length; j++) {
                                        result.push(arr[i][j]);
                                    }
                                }
                                else {
                                    // Just push if anything else
                                    result.push(arr[i]);
                                }
                            }
                        }
                        return [2 /*return*/, result];
                }
            });
        });
    };
    return Flatten;
}());
exports.Flatten = Flatten;
